src/migration/service-agreements-metadata-migration.js

```diff
@@ -0,0 +1,152 @@
+    ATTEMPTED_PROOF_COMMAND_STATUS,
+} from '../constants/constants.js';
+
+class ServiceAgreementsMetadataMigration extends BaseMigration {
+    constructor(
+        migrationName,
+        logger,
+        config,
+        tripleStoreService,
+        blockchainModuleManager,
+        repositoryModuleManager,
+        serviceAgreementService,
+        ualService,
+    ) {
+        super(migrationName, logger, config);
+        this.blockchainModuleManager = blockchainModuleManager;
+        this.repositoryModuleManager = repositoryModuleManager;
+        this.serviceAgreementService = serviceAgreementService;
+        this.ualService = ualService;
+        this.tripleStoreService = tripleStoreService;
+    }
+
+    async executeMigration() {
+        // get metadata of all stored assets in public current triple store
+        const query = `PREFIX schema: <${SCHEMA_CONTEXT}>
+                        SELECT DISTINCT ?ual  WHERE {
+                            GRAPH <assets:graph> {
+                                    ?ual ?p ?o
+                            }
+                        }`;
+        const assetsMetadata = await this.tripleStoreService.select(
+            TRIPLE_STORE_REPOSITORIES.PUBLIC_CURRENT,
+            query,
+        );
+        const identities = {};
+        // for each asset
+        for (const { ual } of assetsMetadata) {
+            const { blockchain, contract, tokenId } = this.ualService.resolveUAL(ual);
+            if (!identities[blockchain]) {
+                // eslint-disable-next-line no-await-in-loop
+                identities[blockchain] = await this.blockchainModuleManager.getIdentityId(
+                    blockchain,
+                );
+            }
+            // eslint-disable-next-line no-await-in-loop
+            await this.updateTables(blockchain, contract, tokenId, identities[blockchain]);
+        }
+    }
+
+    async updateTables(blockchain, contract, tokenId, identityId) {
+        // get assertion ids
+        const assertionIds = await this.blockchainModuleManager.getAssertionIds(
+            blockchain,
+            contract,
+            tokenId,
+        );
+        const stateIndex = assertionIds.length - 1;
+
+        // calculate keyword
+        const keyword = this.blockchainModuleManager.encodePacked(
+            blockchain,
+            ['address', 'bytes32'],
+            [contract, assertionIds[0]],
+        );
+
+        // generate agreement id
+        const agreementId = await this.serviceAgreementService.generateId(
+            blockchain,
+            contract,
+            tokenId,
+            keyword,
+            CONTENT_ASSET_HASH_FUNCTION_ID,
+        );
+
+        // get agreement data
+        const agreementData = await this.blockchainModuleManager.getAgreementData(
+            blockchain,
+            agreementId,
+        );
+
+        // store in service_agreements table
+        await this.repositoryModuleManager.updateServiceAgreementRecord(
```

Comment:
We can optimise this by doing bulk updates and inserts, same thing for attempted commit/proof